### PR TITLE
Disallow conflicting overlapping writes when doing an immutable upload

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3801

Additionally, this defines the equivalent of `RIBucketWriter.abort()` in the HTTP protocol, since if somehow there are conflicting writes the only reasonable thing to do is start from scratch, so that needs to be exposed (plus some existing code assumes aborts are possible.)

The followup, https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3793, will then make BucketWriter persist for more than one Foolscap connection, so that:

1. Multiple Foolscap connections can write (not likely in practice).
2. Parallel HTTP uploads can write.

### CODE REVIEW

File: src/allmydata/test/test_istorageserver.py [hunk 1]
```diff
@@ -248,12 +248,31 @@ def test_written_shares_are_readable(self):
             (yield buckets[2].callRemote("read", 0, 1024)), b"3" * 512 + b"4" * 512
         )
 
-    @skipIf(True, "https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3801")
+    @inlineCallbacks
     def test_overlapping_writes(self):
         """
-        The policy for overlapping writes is TBD:
-        https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3801
+        Overlapping, non-identical writes in immutable uploads fail.
```
Comment (exarkun) [inline src/allmydata/test/test_istorageserver.py:254]: Should we have an interface test for overlapping identical writes too?

File: src/allmydata/storage/immutable.py [hunk 2]
```diff
@@ -226,7 +231,20 @@ def remote_write(self, offset, data):
         precondition(not self.closed)
         if self.throw_out_all_data:
             return
+
+        # Make sure we're not conflicting with existing data:
+        end = offset + len(data)
+        for (chunk_start, chunk_stop, _) in self._already_written.ranges(offset, end):
```
Comment (exarkun) [inline src/allmydata/storage/immutable.py:237]: nice

**Review by exarkun — COMMENTED** [review]
Thanks.  Some comments inline, one about packaging.  I can probably help with that one if you want.  Others are a bit open-ended so let me know if you want to discuss any of those.

File: docs/proposed/http-storage-node-protocol.rst [hunk 3]
```diff
@@ -497,22 +497,26 @@ If any one of these requests fails then at most 128KiB of upload work needs to b
...
-  That is::
-
-    { "required":
-      [ { "begin": <byte position, inclusive>
-        , "end":   <byte position, exclusive>
-        }
-      ,
-      ...
-      ]
-    }
```
[the remark this answers is not in the document]
Comment (itamarst) [reply docs/proposed/http-storage-node-protocol.rst:515]: Sure, makes sense.

File: src/allmydata/test/test_istorageserver.py (hunk 1 above)
Comment (itamarst) [reply src/allmydata/test/test_istorageserver.py:254]: OK.

File: nix/tahoe-lafs.nix [hunk 4]
```diff
@@ -97,7 +97,7 @@ EOF
     setuptoolsTrial pyasn1 zope_interface
     service-identity pyyaml magic-wormhole treq
     eliot autobahn cryptography netifaces setuptools
-    future pyutil distro configparser
+    future pyutil distro configparser collections-extended
```
[the remark this answers is not in the document]
Comment (itamarst) [reply nix/tahoe-lafs.nix:100]: OK will add it to overlay.

File: docs/proposed/http-storage-node-protocol.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (itamarst) [reply docs/proposed/http-storage-node-protocol.rst:515]: Actually, for CREATED it's no longer relevant, and for conflict it's... not helpful. So just for OK I think.

**Comment by itamarst** [comment]
Need to restore my Nix dev environment.

**Comment by itamarst** [comment]
Looks like there was intermittent error due to Tor git site being down or something, rerunning tests.

**Pull request merged**